solr/core/src/java/org/apache/solr/search/facet/FacetFieldProcessorByHashDVString.java

```diff
@@ -0,0 +1,300 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.search.facet;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.function.IntFunction;
+
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.search.SimpleCollector;
+import org.apache.lucene.util.BitUtil;
+import org.apache.lucene.util.BytesRef;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.util.SimpleOrderedMap;
+import org.apache.solr.schema.SchemaField;
+import org.apache.solr.search.DocSetUtil;
+import org.apache.solr.search.facet.SlotAcc.SlotContext;
+
+/**
+ * Facets DocValues into a HashMap using the BytesRef as key.
+ * Limitations:
+ * <ul>
+ *   <li>Only for string type fields not numerics (use dvhash for those)</li>
+ * </ul>
+ */
+class FacetFieldProcessorByHashDVString extends FacetFieldProcessor {
+
+  static class TermData {
+    int count;
+    int slotIndex;
+  }
+
+  // Using a regular HashMap hence slots get created dynamically as new keys are found from docvalues
+  private HashMap<BytesRef, TermData> table;
+  private ArrayList<BytesRef> slotList; // position in List is the slot number, value is key for table
+  private int capacity; // how many slots we will need for accs, gets resized later if needed
+
+  FacetFieldProcessorByHashDVString(FacetContext fcontext, FacetField freq, SchemaField sf) {
+    super(fcontext, freq, sf);
+    if (freq.mincount == 0) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
+          getClass()+" doesn't support mincount=0");
+    }
+    if (freq.prefix != null) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
+          getClass()+" doesn't support prefix"); // yet, but it could
+    }
+    FieldInfo fieldInfo = fcontext.searcher.getSlowAtomicReader().getFieldInfos().fieldInfo(sf.getName());
+    if (fieldInfo != null &&
+        fieldInfo.getDocValuesType() != DocValuesType.SORTED &&
+        fieldInfo.getDocValuesType() != DocValuesType.SORTED_SET) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
+          getClass()+" only supports string with docValues");
+    }
+  }
+
+  @Override
+  public void process() throws IOException {
+    super.process();
+    response = calcFacets();
+    table = null;//gc
+    slotList = null;
+  }
+
+  private SimpleOrderedMap<Object> calcFacets() throws IOException {
+
+    int possibleValues = fcontext.base.size();
+    int hashSize = BitUtil.nextHighestPowerOfTwo((int) (possibleValues * (1 / 0.7) + 1));
+    hashSize = Math.min(hashSize, 1024);
+
+    table = new HashMap<>(hashSize);
+    slotList = new ArrayList<>();
```

Comment:
Does this needs presizing in any way too?

Reply:
It doesn't need to, really. And actually the hashmap size doesn't need to be picked either.. largely this code was started by copying the DVhash class and ripping out their custom hashtable. I could just remove this hashSize calc.